README.md
## @zonst/switch-system
统一系统切换功能-弹出dialog

##使用方式

```js
pnpm install @zonst/switch-system -S

import injectSwitchSystem from '@zonst/switch-system';

// 组件加载完成后init
injectSwitchSystem.init({
  container: document.getElementById('ZonstCommonSwitchSystem')
})
```

```js
删除包 - 先修改package.json中version要删除的版本
然后执行pnpm run unpublish
```

CHANGELOG.md
<!--
 * @Author: lipengcheng
 * @Date: 2023-07-24 09:44:39
 * @LastEditTime: 2023-07-24 09:44:51
 * @Description: 版本变更说明
-->


src/index.js
// import { $http } from './utils/http'
// import './index.scss'
import './base.scss'
import './el-drawer.scss'
import operateDom from './utils/operateDialog'

export default class injectSwitchSystem extends operateDom {
  constructor() {
    super()
  }
  static url = 'https://zlink.zonst.com'

  static openDialog() {
    super.domdialog.style.display = 'block'
    const a = setTimeout(() => {
      super.domdialog.querySelector('.el-drawer').classList.add('open')
      clearTimeout(a)
    }, 50)
    super.domdialog.querySelector('.el-overlay').addEventListener('click', super.closeDialog, false); // 监听点击空白处
    super.keyDownListener() // 监听esc按下
  }

  static async init(dom) {
    console.log(dom)
    try {
      const baseUrl = dom.env ? 'http://zlink.test.zonst.com' : 'https://zlink.zonst.com'
      this.url = baseUrl
      await super.request(this.url) // 请求数据
      super.createdDom() // 创建Dom
      dom.container.addEventListener('click', this.openDialog, false);
    } catch (err) {
      console.log(err)
    }
  }
};


rollup.config.js
/*
 * @Author: lipengcheng
 * @Date: 2023-07-24 09:52:40
 * @LastEditTime: 2023-07-24 16:18:49
 * @Description: 
 */
import json from "@rollup/plugin-json"
import resolve from '@rollup/plugin-node-resolve' // 配置rollup引入node_modules模块，处理外部依赖
import commonjs from '@rollup/plugin-commonjs'  // 支持解析commonjs模块
// import commonjs from "rollup-plugin-commonjs"
import { terser } from 'rollup-plugin-terser' // 代码压缩插件
// import { eslint } from 'rollup-plugin-eslint'
import { babel } from '@rollup/plugin-babel' // 转译成es5
import autoprefixer from 'autoprefixer' // css兼容前缀插件
import cssnano from 'cssnano' // css压缩
import postcss from "rollup-plugin-postcss" // 集成scss\less\styls


export default {
  // input:”src/index.js”,  //指定入口文件路径
  // 多入口打包
  // input:[‘src/index.js’,’src/album.js’],
  input: "src/index.js",
  output: [
    {
      // file:”dist/bundle.js”, // 指定输出的文件名
      // format:”iife”, // 指定输出的格式
      dir: "dist",
      format: 'cjs',
      entryFileNames: '[name].cjs.js',
    },
    {
      dir: "dist",
      format: 'esm',
      entryFileNames: '[name].esm.js',
    },
    {
      dir: "dist",
      format: 'iife',
      entryFileNames: '[name]dev.iife.js',
    }
  ],
  plugins:[
    // eslint(),
    resolve(), commonjs(), terser(), json(),
    babel({ exclude: 'node_modules/**' }), // 不转译，node_modules里面的代码
    postcss({ plugins: [autoprefixer(), cssnano()] }),
  ]
}


src/utils/http.js
/*
 * @Author: lipengcheng
 * @Date: 2023-07-24 10:39:43
 * @LastEditTime: 2023-07-28 16:12:26
 * @Description: http封装
 */

// import axios from 'axios'

const Toast = (msg, duration) => {
  duration = isNaN(duration) ? 3000 : duration;
  var m = document.createElement('div');
  m.innerHTML = msg;
  m.style.cssText="max-width:60%;min-width: 150px;padding:0 14px;height: 40px;color: rgb(255, 255, 255);line-height: 40px;text-align: center;border-radius: 4px;position: fixed;top: 50%;left: 50%;transform: translate(-50%, -50%);z-index: 9999999999;background: rgba(0, 0, 0,.7);font-size: 12px;";
  document.body.appendChild(m);
  setTimeout(function() {
    var d = 0.5;
    m.style.webkitTransition = '-webkit-transform ' + d + 's ease-in, opacity ' + d + 's ease-in';
    m.style.opacity = '0';
    setTimeout(function() { document.body.removeChild(m) }, d * 1000);
  }, duration);
}

const $http = (method, url, data) => {
  return new Promise((resolve, reject) => {
    const headers = {}
    const dataOptions = {}
    if (method === 'get') {
      dataOptions.params = data
    } else {
      dataOptions.data = data
    }
    const token = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VyX2lkIjoxLCJ1c2VyX25hbWUiOiLmtYvor5VhZG1pbiIsImdyb3VwX2lkIjo3LCJpc19zdXBlcnVzZXIiOmZhbHNlLCJleHAiOjE2OTA4NTMxNDd9.bo1O8Wo3A-qVP_n8z9vJjFJ6Qd3XIyLe9qWZjsK7pm4' // localStorage.getItem('ZLINK_TOKEN_NAME')
    headers['X-Xq5-Jwt'] = token ?? ''
    
    const xhr = new XMLHttpRequest();
    xhr.open(method, url, true);
    console.log(headers)
    xhr.setRequestHeader('X-Xq5-Jwt', token)
    xhr.send();
    xhr.onreadystatechange = function() {
      if (xhr.readyState == 4) {
        if (xhr.status >= 200 && xhr.status < 300) {
          resolve(xhr.response);
        } else {
          reject(xhr.status);
          Toast('请求错误!', 2000);
        }
      }
    }
    /* axios({
      url,
      method,
      timeout: 10000,
      headers,
      ...dataOptions,
    }).then(async (res) => {
      if (res.status === 200) {
        if (res.data.code === 0) {
          resolve(res.data)
        } else {
          reject(res)
          Toast('请求错误!', 2000);
        }
      } else {
        reject(res)
        Toast('请求错误!', 2000);
      }
    }) */
  })
}

export { $http }


rollup.config.dev.js
/*
 * @Author: lipengcheng
 * @Date: 2023-07-24 09:52:40
 * @LastEditTime: 2023-07-25 17:06:34
 * @Description: 
 */
import json from "@rollup/plugin-json"
import resolve from '@rollup/plugin-node-resolve' // 配置rollup引入node_modules模块，处理外部依赖
import commonjs from '@rollup/plugin-commonjs'  // 支持解析commonjs模块
// import commonjs from "rollup-plugin-commonjs"
import { terser } from 'rollup-plugin-terser' // 代码压缩插件
// import { eslint } from 'rollup-plugin-eslint'
import { babel } from '@rollup/plugin-babel' // 转译成es5
import autoprefixer from 'autoprefixer' // css兼容前缀插件
import cssnano from 'cssnano' // css压缩
import postcss from "rollup-plugin-postcss" // 集成scss\less\styls

import serve from "rollup-plugin-serve"
import livereload from 'rollup-plugin-livereload'

export default {
  // input:”src/index.js”,  //指定入口文件路径
  // 多入口打包
  // input:[‘src/index.js’,’src/album.js’],
  input: "src/index.js",
  output: [
    {
      dir: "dist",
      format: 'iife',
      entryFileNames: '[name]dev.iife.js',
      name: 'injectSwitchSystem' // If you do not supply "output.name", you may not be able to access the exports of an IIFE bundle.
    }
  ],
  plugins:[
    // eslint(),
    resolve({
      browser: true // 如果打包后的代码如果是使用在浏览器上的，需要进行配置
    }),
    commonjs(), json(),
    babel({ babelHelpers: 'bundled', exclude: 'node_modules/**' }), // 不转译，node_modules里面的代码
    postcss({ plugins: [autoprefixer(), cssnano()] }),
    serve({ // 打开html
      // open: true,
      contentBase: "./",
    }),
    livereload(),
  ]
}


src/utils/createdDom.js
import { $http } from './http'

const dom = `
    <div role="dialog" aria-modal="true" aria-label="系统切换" class="el-overlay-dialog">
      <div class="el-dialog" tabindex="-1">
        <header class="el-dialog__header">
          <span role="heading" class="el-dialog__title">系统切换</span>
          <button aria-label="Close this dialog" class="el-dialog__headerbtn" type="button">
            <i class="el-icon el-dialog__close">
              <svg viewBox="0 0 1024 1024" xmlns="http://www.w3.org/2000/svg">
                <path fill="currentColor" d="M764.288 214.592 512 466.88 259.712 214.592a31.936 31.936 0 0 0-45.12 45.12L466.752 512 214.528 764.224a31.936 31.936 0 1 0 45.12 45.184L512 557.184l252.288 252.288a31.936 31.936 0 0 0 45.12-45.12L557.12 512.064l252.288-252.352a31.936 31.936 0 1 0-45.12-45.184z">
                </path>
              </svg>
            </i>
          </button>
        </header>
        <div class="el-dialog__body">

          <form class="el-form">
            <div class="el-form-item">
              <label id="el-id-9829-351" for="el-id-9829-405" class="el-form-item__label" style="">系统选择：</label>
              <div class="el-form-item__content">
                <div class="el-input">
                  <div class="el-input__wrapper" role="button" tabindex="-1">
                    <select class="el-input__inner" autocomplete="off" tabindex="0" id="el-id-9829-405">
                      <option :value="item.value" class="option_list">
                        {{ item.label }}
                      </option>
                      <option :value="item.value" class="option_list">
                        {{ item.label }}
                      </option>
                      <option :value="item.value" class="option_list">
                        {{ item.label }}
                      </option>
                    </select>
                  </div>
                </div>
              </div>
            </div>
            <div class="el-form-item">
              <label id="el-id-9829-352" for="el-id-9829-406" class="el-form-item__label" style="">环境选择：</label>
              <div class="el-form-item__content">
                <div class="el-input">
                  <div class="el-input__wrapper" role="button" tabindex="-1">
                    <select class="el-input__inner" autocomplete="off" tabindex="0" id="el-id-9829-406">
                      <option :value="item.value" class="option_list">
                        {{ item.label }}
                      </option>
                      <option :value="item.value" class="option_list">
                        {{ item.label }}
                      </option>
                      <option :value="item.value" class="option_list">
                        {{ item.label }}
                      </option>
                    </select>
                  </div>
                </div>
              </div>
            </div>
          </form>

        </div>
        <footer class="el-dialog__footer">
          <span class="dialog-footer">
            <button aria-disabled="false" type="button" class="el-button">
              <span class="">取消</span>
            </button>
            <button aria-disabled="false" type="button" class="el-button el-button--primary">
              <span class=""> 确定 </span>
            </button>
          </span>
        </footer>
      </div>
    </div>
`

console.log($http)
$http('get', 'http://zlink.test.xq5.com/api/v1/platforms/all').then(res => {
  
})
const domdialog = document.createElement('div')
domdialog.className = 'zonst-switch-system'
domdialog.style.zIndex = '2099'
domdialog.innerHTML = dom
document.body.appendChild(domdialog)

export default domdialog


src/utils/operateDialog.js
/*
 * @Author: lipengcheng
 * @Date: 2023-07-24 17:15:28
 * @LastEditTime: 2023-07-28 18:05:38
 * @Description: 生成dialog dom及操作
 */
import { $http } from './http'

class operateDom {
  constructor() {}
  static systemData = []
  
  static systemHTML = '<option value="" class="option_list"> 请选择系统 </option>'
  static envData = []
  static envHTML = '<option value="" class="option_list"> 请选择环境 </option>'

  static domdialog = document.createElement('div')

  static drawerDom = `
    <div class="el-overlay" style="z-index: 2039;">
      <div aria-modal="true" aria-label="系统切换" aria-describedby="el-id-2609-193" class="el-drawer rtl" role="dialog" style="width: 80%;">
        <span class="el-drawer__sr-focus" tabindex="-1"></span>
        <header class="el-drawer__header">
          <span id="el-id-2609-194" role="heading" class="el-drawer__title">系统切换</span>
          <button aria-label="Close this dialog" class="el-drawer__close-btn" type="button" onClick="injectSwitchSystem.closeDialog()">
            <i class="el-icon el-drawer__close">
              <svg viewBox="0 0 1024 1024" xmlns="http://www.w3.org/2000/svg">
                <path fill="currentColor" d="M764.288 214.592 512 466.88 259.712 214.592a31.936 31.936 0 0 0-45.12 45.12L466.752 512 214.528 764.224a31.936 31.936 0 1 0 45.12 45.184L512 557.184l252.288 252.288a31.936 31.936 0 0 0 45.12-45.12L557.12 512.064l252.288-252.352a31.936 31.936 0 1 0-45.12-45.184z"></path>
              </svg>
            </i>
          </button>
        </header>
        <div id="el-id-2609-193" class="el-drawer__body">
        </div>
      </div>
    </div>`
  
/*   static dom = `
    <div role="dialog" aria-modal="true" aria-label="系统切换" class="el-overlay-dialog">
      <div class="el-dialog" tabindex="-1">
        <header class="el-dialog__header">
          <span role="heading" class="el-dialog__title">系统切换</span>
          <button aria-label="Close this dialog" class="el-dialog__headerbtn" type="button" onClick="${ operateDom.closeDialog }">
            <i class="el-icon el-dialog__close">
              <svg viewBox="0 0 1024 1024" xmlns="http://www.w3.org/2000/svg">
                <path fill="currentColor" d="M764.288 214.592 512 466.88 259.712 214.592a31.936 31.936 0 0 0-45.12 45.12L466.752 512 214.528 764.224a31.936 31.936 0 1 0 45.12 45.184L512 557.184l252.288 252.288a31.936 31.936 0 0 0 45.12-45.12L557.12 512.064l252.288-252.352a31.936 31.936 0 1 0-45.12-45.184z">
                </path>
              </svg>
            </i>
          </button>
        </header>
        <div class="el-dialog__body">

          <form class="el-form">
            <div class="el-form-item">
              <label id="el-id-9829-351" for="el-id-9829-405" class="el-form-item__label" style="">系统选择：</label>
              <div class="el-form-item__content">
                <div class="el-input">
                  <div class="el-input__wrapper" role="button" tabindex="-1">
                    <select class="el-input__inner select_system_container" autocomplete="off" tabindex="0" id="el-id-9829-405" onChange="injectSwitchSystem.systemChange(this.options[this.options.selectedIndex].value)">
                      `+ this.systemHTML + `
                    </select>
                  </div>
                </div>
              </div>
            </div>
            <div class="el-form-item env-form-item">
              <label id="el-id-9829-352" for="el-id-9829-406" class="el-form-item__label" style="">环境选择：</label>
              <div class="el-form-item__content">
                <div class="el-input">
                  <div class="el-input__wrapper" role="button" tabindex="-1">
                    <select class="el-input__inner select_env_container" autocomplete="off" tabindex="0" id="el-id-9829-406" onChange="injectSwitchSystem.systemChange(this.options[this.options.selectedIndex].value)">
                      `+ this.envHTML + `
                    </select>
                  </div>
                </div>
              </div>
            </div>
          </form>

        </div>
        <footer class="el-dialog__footer">
          <span class="dialog-footer">
            <button aria-disabled="false" type="button" class="el-button" onClick="injectSwitchSystem.closeDialog()">
              <span class="">取消</span>
            </button>
            <button aria-disabled="false" type="button" class="el-button el-button--primary">
              <span class=""> 确定 </span>
            </button>
          </span>
        </footer>
      </div>
    </div>
  ` */

  static async request(base) {
    await $http('GET', base + '/api/server/v1/permission/account?platformKey=zlink').then(res => {
      // this.systemData = res.data?.list || []
      const data = {
        operation: {
          name: '运营系统',
          list: [
            { id: 1, name: '运营系统1', url: 'www.baidu.com', isEnv: true,
              children: [
                { id: 11, name: '运营系统1-1', url: 'www.baidu.com' },
                { id: 12, name: '运营系统1-2', url: 'www.baidu.com' },
                { id: 13, name: '运营系统1-3', url: 'www.baidu.com' },
              ]
            },
            { id: 2, name: '运营系统2', url: 'www.baidu.com', isEnv: false },
            { id: 3, name: '运营系统3', url: 'www.baidu.com', isEnv: false },
            { id: 4, name: '运营系统4', url: 'www.baidu.com', isEnv: false },
            { id: 5, name: '运营系统5', url: 'www.baidu.com', isEnv: true,
              children: [
                { id: 51, name: '运营系统5-1', url: 'www.baidu.com' },
                { id: 52, name: '运营系统5-2', url: 'www.baidu.com' },
                { id: 53, name: '运营系统5-3', url: 'www.baidu.com' },
              ]
            }
          ]
        },
        science: {
          name: '技术平台',
          list: [
            { id: 1, name: '技术平台1', url: 'www.baidu.com', isEnv: true,
              children: [
                { id: 11, name: '技术平台1-1', url: 'www.baidu.com' },
                { id: 12, name: '技术平台1-2', url: 'www.baidu.com' },
                { id: 13, name: '技术平台1-3', url: 'www.baidu.com' },
              ]
            },
            { id: 2, name: '技术平台2', url: 'www.baidu.com', isEnv: false },
            { id: 3, name: '技术平台3', url: 'www.baidu.com', isEnv: false },
            { id: 4, name: '技术平台4', url: 'www.baidu.com', isEnv: false },
            { id: 5, name: '技术平台5', url: 'www.baidu.com', isEnv: true,
              children: [
                { id: 51, name: '技术平台5-1', url: 'www.baidu.com' },
                { id: 52, name: '技术平台5-2', url: 'www.baidu.com' },
                { id: 53, name: '技术平台5-3', url: 'www.baidu.com' },
              ]
            }
          ]
        }
      }

      let list = ''
      for (let o in data) {
        console.log(o)
        list += `<div class="lpc-classify-container">
          <div>${ data[o].name }</div>` + 
          data[o].list.forEach((l) => {
            return `<div class="platform-list"></div>`
          })
       + `</div>`


        
        // list += `<div class="platform_list"> ${ o.name } </div>`
      }
      console.log(list)







      
      /* this.systemData = [
        { id: 1, name: '运营系统1', url: 'www.baidu.com', isEnv: true,
          children: [
            { id: 11, name: '运营系统1-1', url: 'www.baidu.com' },
            { id: 12, name: '运营系统1-2', url: 'www.baidu.com' },
            { id: 13, name: '运营系统1-3', url: 'www.baidu.com' },
          ]
        },
        { id: 2, name: '运营系统2', url: 'www.baidu.com', isEnv: false },
        { id: 3, name: '运营系统3', url: 'www.baidu.com', isEnv: false },
        { id: 4, name: '运营系统4', url: 'www.baidu.com', isEnv: false },
        { id: 5, name: '运营系统5', url: 'www.baidu.com', isEnv: true,
          children: [
            { id: 51, name: '运营系统5-1', url: 'www.baidu.com' },
            { id: 52, name: '运营系统5-2', url: 'www.baidu.com' },
            { id: 53, name: '运营系统5-3', url: 'www.baidu.com' },
          ]
        }
      ]
      for (let i = 0; i < this.systemData.length; i++) {
        this.systemHTML += `<option value="${ this.systemData[i].id }" class="option_list"> ${ this.systemData[i].name } </option>`
      }
      if (this.systemData[0].isEnv) { // 如果第一个有环境变量
        this.envData = this.systemData[0].children
        this.envData.forEach((opt) => {
          this.envHTML += `<option value="${ opt.id }" class="option_list"> ${ opt.name } </option>`
        })
      } */
    })
  }

  static createdDom() {
    this.domdialog.className = 'zonst-switch-system'
    this.domdialog.style.zIndex = '2099'
    this.domdialog.innerHTML = this.drawerDom // this.dom
    this.domdialog.style.display = 'none'
    document.body.appendChild(this.domdialog)
    // document.querySelector('.select_system_container').innerHTML = this.systemHTML
    // document.querySelector('.select_env_container').innerHTML = this.envHTML
  }

  static close() { 
    document.querySelector('.zonst-switch-system .el-drawer').classList.remove('open')
    const a = setTimeout(() => {
      document.querySelector('.zonst-switch-system').style.display = 'none'
      clearTimeout(a)
    }, 500)
  }

  static closeDialog(e) {
    if (e) {
      if ((typeof e.target.className) !== 'object' && e.target.className.indexOf('el-overlay') !== -1) {
        console.log(this)
        operateDom.close();
      }
      return
    }
    this.close();
  }

  static keyDownListener() {
    document.onkeydown = function (event) {
      const e = event || window.event || arguments.callee.caller.arguments[0];
      if (e && e.keyCode == 27) { // 按 Esc
        if (document.querySelector('.zonst-switch-system').style.display !== 'block') return
        operateDom.close()
      }
    }
  }

  static systemChange(v) {
    console.log(v)
    if (!v) return
    this.systemData.forEach(list => {
      if (list.id === +v) {
        if (!list.isEnv) {
          document.querySelector('.env-form-item').style.display = 'none'
        } else {
          this.envData = list.children
          this.envHTML = '<option value="" class="option_list"> 请选择环境 </option>'
          this.envData.forEach((opt) => {
            this.envHTML += `<option value="${ opt.id }" class="option_list"> ${ opt.name } </option>`
          })
          document.querySelector('.select_env_container').innerHTML = this.envHTML
          document.querySelector('.env-form-item').style.display = 'block'
        }
      }
    })
  }
}

export default operateDom
